File: shore-daemon/src/tools/activity.rs
use super::{ToolCategory, ToolContext, ToolDef, ToolError};
use serde_json::{json, Value};

// ---------------------------------------------------------------------------
// Tool definitions
// ---------------------------------------------------------------------------

pub fn tool_defs() -> Vec<ToolDef> {
    vec![ToolDef {
        name: "activity_heatmap",
        description: "Show the user's message activity patterns as a heatmap by hour of day.",
        parameters: json!({
            "type": "object",
            "properties": {
                "days": {
                    "type": "integer",
                    "description": "Number of days of history to include.",
                    "default": 30
                }
            }
        }),
        category: ToolCategory::Other,
    }]
}

// ---------------------------------------------------------------------------
// Handler
// ---------------------------------------------------------------------------

/// Handle `activity_heatmap` — returns real data from the ActivityTracker
/// when available, otherwise an empty heatmap.
pub async fn handle_activity_heatmap(
    input: Value,
    ctx: &dyn ToolContext,
) -> Result<Value, ToolError> {
    let days = input
        .get("days")
        .and_then(|v| v.as_u64())
        .unwrap_or(30);

    let character = ctx.character_name();
    let autonomy = ctx.autonomy_manager();

    let stats_opt = autonomy.and_then(|mgr| mgr.activity_stats(character));

    match stats_opt {
        Some((stats, message_count)) => {
            let hours: Vec<Value> = (0..24)
                .map(|h| {
                    let class = match stats.hour_classifications[h] {
                        crate::autonomy::activity::HourClassification::Peak => "peak",
                        crate::autonomy::activity::HourClassification::Trough => "trough",
                        crate::autonomy::activity::HourClassification::Normal => "normal",
                    };
                    json!({
                        "hour": h,
                        "density": stats.hour_histogram[h],
                        "classification": class,
                    })
                })
                .collect();

            Ok(json!({
                "days": days,
                "hours": hours,
                "total_messages": message_count,
                "has_sufficient_data": stats.has_sufficient_heatmap,
                "engagement_score": stats.engagement_score,
                "sessions_per_day": stats.sessions_per_day,
            }))
        }
        None => {
            // No autonomy data available — return empty heatmap.
            Ok(json!({
                "days": days,
                "hours": (0..24).map(|h| json!({ "hour": h, "density": 0.0, "classification": "normal" })).collect::<Vec<_>>(),
                "total_messages": 0,
                "has_sufficient_data": false,
                "engagement_score": 0.0,
                "sessions_per_day": 0.0,
            }))
        }
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::TestToolContext;

    #[test]
    fn test_activity_tool_defs() {
        let defs = tool_defs();
        assert_eq!(defs.len(), 1);
        assert_eq!(defs[0].name, "activity_heatmap");
        assert_eq!(defs[0].category, ToolCategory::Other);
    }

    #[tokio::test]
    async fn test_activity_heatmap_no_autonomy() {
        // TestToolContext has no autonomy manager — should return empty heatmap.
        let ctx = TestToolContext::new();
        let result = handle_activity_heatmap(json!({}), &ctx).await.unwrap();
        assert_eq!(result["days"], 30);
        assert_eq!(result["total_messages"], 0);
        assert_eq!(result["has_sufficient_data"], false);

        let hours = result["hours"].as_array().unwrap();
        assert_eq!(hours.len(), 24);
    }

    #[tokio::test]
    async fn test_activity_heatmap_custom_days() {
        let ctx = TestToolContext::new();
        let result = handle_activity_heatmap(json!({"days": 7}), &ctx).await.unwrap();
        assert_eq!(result["days"], 7);
    }
}

File: backend/daemon-ts/src/tools/activity.ts
/**
 * `activity_heatmap` — returns real autonomy data when a stats hook is
 * wired into the context, otherwise an empty 24-hour heatmap.
 *
 * Ported from `backend/daemon/src/tools/activity.rs`. The autonomy
 * subsystem itself ships in Phase 8 — until then `ctx.activityStats`
 * is always undefined and this tool returns the empty shape.
 */

import type { ToolContext, ToolHandler } from "./registry.ts";

export const ACTIVITY_HEATMAP_DESCRIPTION =
  "View {{user}}'s activity heatmap — when they typically message you, broken down by hour of day and day of week. Use when you want to understand their rhythm: whether now is an unusual time for them to be talking to you, when they're most likely to be around, or whether they've been more or less active lately. Returns hour-by-hour density, classification (peak/trough/normal), total user-turn count over the window, and an engagement score. The `days` parameter controls how much history is included; 30 is a good default for seeing a steady pattern.";

function emptyHeatmap(days: number): unknown {
  const hours = [];
  for (let h = 0; h < 24; h++) {
    hours.push({ hour: h, density: 0.0, classification: "normal" });
  }
  return {
    days,
    hours,
    total_messages: 0,
    total_turns: 0,
    has_sufficient_data: false,
    engagement_score: 0.0,
    sessions_per_day: 0.0,
  };
}

export const activityHeatmapHandler: ToolHandler = {
  name: "activity_heatmap",
  description: ACTIVITY_HEATMAP_DESCRIPTION,
  inputSchema: {
    type: "object",
    properties: {
      days: {
        type: "integer",
        description: "Number of days of history to include.",
        default: 30,
      },
    },
  },
  async execute(input: unknown, ctx: ToolContext): Promise<string> {
    const obj = (input ?? {}) as { days?: unknown };
    const days =
      typeof obj.days === "number" && Number.isFinite(obj.days)
        ? Math.max(0, Math.floor(obj.days))
        : 30;

    const stats = ctx.activityStats?.(ctx.characterName);
    if (stats === undefined) return JSON.stringify(emptyHeatmap(days));

    const hours = [];
    for (let h = 0; h < 24; h++) {
      hours.push({
        hour: h,
        density: stats.hourHistogram[h] ?? 0,
        classification: stats.hourClassifications[h] ?? "normal",
      });
    }
    return JSON.stringify({
      days,
      hours,
      total_messages: stats.turnCount,
      total_turns: stats.turnCount,
      has_sufficient_data: stats.hasSufficientHeatmap,
      engagement_score: stats.engagementScore,
      sessions_per_day: stats.sessionsPerDay,
    });
  },
};
